File: objc/Project2/Project2 WatchKit Extension/InterfaceController.m
//
//  InterfaceController.m
//  Project2 WatchKit Extension
//
//  Created by Jinwoo Kim on 1/22/21.
//

#import "InterfaceController.h"

@implementation InterfaceController

- (void)setup {
    self.allMoves = [@[@"rock", @"paper", @"scissors"] mutableCopy];
    self.shouldWin = YES;
    self.level = 1;
}

- (void)awakeWithContext:(id)context {
    [self setup];
    [super awakeWithContext:context];
    
    [self.rock setBackgroundImage:[UIImage imageNamed:@"rock"]];
    [self.paper setBackgroundImage:[UIImage imageNamed:@"paper"]];
    [self.scissors setBackgroundImage:[UIImage imageNamed:@"scissors"]];
    
    [self.timer start];
    [self newLevel];
}

- (void)newLevel {
    if (self.level == 21) {
        [self.result setHidden:NO];
        [self.question setHidden:YES];
        [self.answers setHidden:YES];
        [self.timer stop];
        return;
    }
    
    [self.levelCounter setText:[NSString stringWithFormat:@"%d/20", self.level]];
    BOOL randomBool = arc4random_uniform(2) == 1;
    
    if (randomBool) {
        [self setTitle:@"Win!"];
        self.shouldWin = YES;
    } else {
        [self setTitle:@"Lose!"];
        self.shouldWin = NO;
    }
    
    [self.allMoves shuffle];
    [self.question setImage:[UIImage imageNamed:self.allMoves[0]]];
}

- (void)checkFor:(NSString *)answer {
    if ([self.allMoves[0] isEqual:answer]) {
        self.level += 1;
        [self newLevel];
    } else {
        self.level -= 1;
        if (self.level < 1) self.level = 1;
        [self newLevel];
    }
}

- (void)willActivate {
    // This method is called when watch view controller is about to be visible to user
}

- (void)didDeactivate {
    // This method is called when watch view controller is no longer visible
}

- (IBAction)rockTapped {
    if (self.shouldWin) {
        [self checkFor:@"scissors"];
    } else {
        [self checkFor:@"paper"];
    }
}

- (IBAction)paperTapped {
    if (self.shouldWin) {
        [self checkFor:@"rock"];
    } else {
        [self checkFor:@"scissors"];
    }
}

- (IBAction)scissorsTapped {
    if (self.shouldWin) {
        [self checkFor:@"paper"];
    } else {
        [self checkFor:@"rock"];
    }
}


@end

File: swift/Project2/Project2 WatchKit Extension/InterfaceController.swift
//
//  InterfaceController.swift
//  Project2 WatchKit Extension
//
//  Created by Jinwoo Kim on 1/22/21.
//

import WatchKit
import Foundation


class InterfaceController: WKInterfaceController {
    @IBOutlet weak var result: WKInterfaceLabel!
    @IBOutlet weak var question: WKInterfaceImage!
    @IBOutlet weak var answers: WKInterfaceGroup!
    @IBOutlet weak var rock: WKInterfaceButton!
    @IBOutlet weak var paper: WKInterfaceButton!
    @IBOutlet weak var scissors: WKInterfaceButton!
    @IBOutlet weak var levelCounter: WKInterfaceLabel!
    @IBOutlet weak var timer: WKInterfaceTimer!
    
    var allMoves = ["rock", "paper", "scissors"]
    var shouldWin = true
    var level = 1
    
    override func awake(withContext context: Any?) {
        super.awake(withContext: context)
        
        rock.setBackgroundImage(UIImage(named: "rock"))
        paper.setBackgroundImage(UIImage(named: "paper"))
        scissors.setBackgroundImage(UIImage(named: "scissors"))
        
        timer.start()
        newLevel()
    }
    
    func newLevel() {
        if level == 21 {
            result.setHidden(false)
            question.setHidden(true)
            answers.setHidden(true)
            timer.stop()
            return
        }
        
        levelCounter.setText("\(level)/20")
        
        if Bool.random() {
            setTitle("Win!")
            shouldWin = true
        } else {
            setTitle("Lose!")
            shouldWin = false
        }
        
        allMoves.shuffle()
        question.setImage(UIImage(named: allMoves[0]))
    }
    
    func check(for answer: String) {
        if allMoves[0] == answer {
            level += 1
            newLevel()
        } else {
            level -= 1
            if level < 1 { level = 1 }
            newLevel()
        }
    }
    
    override func willActivate() {
        // This method is called when watch view controller is about to be visible to user
    }
    
    override func didDeactivate() {
        // This method is called when watch view controller is no longer visible
    }
    
    @IBAction func rockTapped() {
        if shouldWin {
            check(for: "scissors")
        } else {
            check(for: "paper")
        }
    }
    
    @IBAction func paperTapped() {
        if shouldWin {
            check(for: "rock")
        } else {
            check(for: "scissors")
        }
    }
    
    @IBAction func scissorsTapped() {
        if shouldWin {
            check(for: "paper")
        } else {
            check(for: "rock")
        }
    }
}
